design-proposals/sidecar-hooks.md

```diff
@@ -0,0 +1,56 @@
+# Sidecar hooks
+
+## Overview
+
+A hooking mechanism that allows customizations to the VMI before it starts without changing KubeVirt
+core components. The common user of those hooks are Sidecars that will run in the same Pod that the
+VMI will be deployed.
+
+## Motivation
+
+A way to accommodate use cases that we might not want to bring and support in KubeVirt itself or to
+give room to test and develop features while they are still not mature enough.
+
+## Goals
+
+- Provide and maintain a set of gRPC APIs for users to do modifications to the VMI
+- Provide and maintain sidecars that are important for core functionalities
+
+## Non Goals
+
+- Provide and maintain a wide variety of sidecars that are not part of KubeVirt's core
+  functionalities.
+
+## User Stories
+
+- As a VM owner, I want to apply custom configurations to QEMU command line
+  - For debugability, by adding support or tweak [to run with debug tools][]
+  - To test unsupported features or apply changes that [could workaround bugs][]
+- As a KubeVirt developer, I want to provide alternative solutions that might not be ready to core
+  KubeVirt components
+  - For Network configurations like [Slirp][] and [Passt][]
+
+## Design
+
+KubeVirt should provide a versioned gRPC Hook module, with all the APIs that can be used including
+parameters and responses.
+
+The gRPC server runs in the Sidecar. The Sidecars will use the module to establish the communication
+with virt-launcher and exchange capabilities to define which Hook version is being used and what are
+the hooks that are implemented.
+
+Virt-launcher will be entitled with managing the hooks. It'll verify and connect to all the Sidecars
+that are running by checking the pre-defined folder where the unix sockets for gRPC communication
+will be created.
+
+Virt-launcher should apply the requested changes to the VMI.
+
+## Hooks API promotion
```

Comment:
I think we are missing the API definition with all its details.
There is a need to express all existing versions and describe what each functionality does.

There is also a need to explain what are the implications and considerations of a sidecar author regarding backward and forward compatibility. Perhaps starting with the virt-launcher side, which relates to the Kubevirt support of hooks and what are the expectations from the server side.

It will be useful to elaborate on what happens when Kubevirt upgrades occur and what happens when the sidecar is upgraded.
Migration also needs to be taken into account, to explain the implications.

Reply:
Thanks, I was in doubt about how deep I should go with the details. I'll go a little bit deeper.